cloudinit/config/cc_kernel_modules.py

```diff
@@ -0,0 +1,396 @@
+
+    @raises RuntimeError
+    """
+
+    try:
+        out = cloud.distro.manage_service("restart", "systemd-modules-load")
+        if re.search("Failed", out.stdout.strip()):
+            raise Exception(out.stdout.strip())
+    except (subp.ProcessExecutionError, Exception) as e:
+        raise RuntimeError(
+            f"Could not load modules with systemd-modules-load:{NL}{str(e)}"
+        ) from e
+
+
+def is_loaded(module_name: str) -> bool:
+    """
+    Checks if a kernel module is already loaded
+
+    @param module_name: name of kernel module
+
+    @return bool
+    """
+    loaded = False
+    try:
+        out = subp.subp(
+            DEFAULT_CONFIG["km_cmd"]["is_loaded"], capture=True, shell=True
+        )
+        if re.search("^" + module_name, out.stdout.strip()):
+            LOG.debug("Kernel module %s is loaded", module_name)
+            loaded = True
+    except subp.ProcessExecutionError as e:
+        util.logexc(
+            LOG,
+            f"Could not determine status of module {module_name}:{NL}{str(e)}",
+        )
+
+    return loaded
+
+
+def unload(modules: list):
+    """Unloads a list of kernel module
+
+    This function unloads a list of kernel modules.
+
+    @param distro_cfg: Distro config
+    @param modules: list of module names
+
+    @raises RuntimeError
+    """
+
+    cmd = copy.copy(DEFAULT_CONFIG["km_cmd"]["unload"])
+    for module in set(modules):
+        cmd.append(module)
+        try:
+            if is_loaded(module):
+                LOG.debug("Unloading kernel module %s", module)
+                subp.subp(cmd)
+        except subp.ProcessExecutionError as e:
+            raise RuntimeError(
+                f"Could not unload kernel module {module}:{NL}{str(e)}"
+            ) from e
+
+
+def update_initial_ramdisk():
+    """Update initramfs for all installed kernels
+
+    @param distro_cfg: Distro config
+
+    @raises RuntimeError
+    """
+    try:
+        subp.subp(DEFAULT_CONFIG["km_cmd"]["update"])
+    except subp.ProcessExecutionError as e:
+        raise RuntimeError(
+            f"Failed to update initial ramdisk:{NL}{str(e)}"
+        ) from e
+
+
+def handle(
+    name: str, cfg: Config, cloud: Cloud, log: Logger, args: list
+) -> None:
+    kernel_modules_section = None
+    unload_modules = []  # type: list
+
+    kernel_modules_section = util.get_cfg_option_list(
+        cfg, "kernel_modules", []
+    )
+
+    if not kernel_modules_section:
+        LOG.debug(
+            "Skipping module named %s, no "
+            "'kernel_modules' configuration found",
+            name,
+        )
+        return
+
+    LOG.debug("Found kernel_modules section in config")
+
+    # check systemd usage
+    if not cloud.distro.uses_systemd():
+        LOG.debug(
+            "Skipping module named %s, due to " "no systemd installed",
+            name,
+        )
+        return
+
+    # cleanup
+    LOG.info("Cleaning up kernel modules")
```

Comment:
This should be LOG.debug and the log logic should live inside cleanup() function